p2p/starknet/block_bodies.go

```diff
@@ -0,0 +1,164 @@
+	"github.com/NethermindEth/juno/utils"
+	"google.golang.org/protobuf/proto"
+)
+
+type streamState int
+
+const (
+	_ streamState = iota
+
+	sendDiff // initial
+	sendClasses
+	sendBlockFin
+	terminal // final state
+)
+
+type blockBodyIterator struct {
+	bcReader blockchain.Reader
+	log      utils.Logger
+
+	state       streamState
+	header      *core.Header
+	stateUpdate *core.StateUpdate
+}
+
+func newBlockBodyIterator(bcReader blockchain.Reader, header *core.Header, log utils.Logger) *blockBodyIterator {
+	return &blockBodyIterator{
+		state:       sendDiff,
+		header:      header,
+		bcReader:    bcReader,
+		log:         log,
+		stateUpdate: nil, // to be filled during one of next states
+	}
+}
+
+func (b *blockBodyIterator) hasNext() bool {
+	statesHaveTransitions := []streamState{
+		sendDiff,
+		sendClasses,
+		sendBlockFin,
+	}
+	return slices.Contains(statesHaveTransitions, b.state)
+}
+
+// Either BlockBodiesResponse_Diff, *_Classes, *_Proof, *_Fin
+func (b *blockBodyIterator) next() (msg proto.Message, valid bool) {
+	switch b.state {
+	case sendDiff:
+		msg, valid = b.diff()
+		b.state = sendClasses
+	case sendClasses:
+		msg, valid = b.classes()
+		b.state = sendBlockFin
+	case sendBlockFin:
+		msg, valid = b.fin()
+		b.state = terminal
+	case terminal:
+		panic("next called on terminal state")
+	default:
+		b.log.Errorw("Unknown state in blockBodyIterator", "state", b.state)
+	}
+
+	return
+}
+
+func (b *blockBodyIterator) classes() (proto.Message, bool) {
+	var classes []*spec.Class
+
+	state, closer, err := b.bcReader.StateAtBlockNumber(b.header.Number)
+	if err != nil {
+		return b.finAndSetTerminalState()
+	}
+	defer b.callAndLogErr(closer, "Error closing state reader in blockBodyIterator.classes()")
+
+	stateDiff := b.stateUpdate.StateDiff
+
+	for _, hash := range stateDiff.DeclaredV0Classes {
+		cls, err := state.Class(hash)
+		if err != nil {
+			return b.finAndSetTerminalState()
+		}
+
+		classes = append(classes, core2p2p.AdaptClass(cls.Class, hash))
+	}
+	for _, class := range stateDiff.DeclaredV1Classes {
+		cls, err := state.Class(class.ClassHash)
+		if err != nil {
+			return b.finAndSetTerminalState()
+		}
+
+		cairo1Cls := cls.Class.(*core.Cairo1Class)
+		classes = append(classes, core2p2p.AdaptClass(cls.Class, cairo1Cls.Hash()))
+	}
+
+	return &spec.BlockBodiesResponse{
+		Id: core2p2p.AdaptBlockID(b.header),
+		BodyMessage: &spec.BlockBodiesResponse_Classes{
+			Classes: &spec.Classes{
+				Domain:  0,
+				Classes: classes,
+			},
+		},
+	}, true
+}
+
+func (b *blockBodyIterator) diff() (proto.Message, bool) {
+	var err error
+	b.stateUpdate, err = b.bcReader.StateUpdateByNumber(b.header.Number)
+	if err != nil {
+		b.log.Errorw("Failed to get state", "err", err)
+		return b.finAndSetTerminalState()
+	}
+	diff := b.stateUpdate.StateDiff
+
+	var contractDiffs []*spec.StateDiff_ContractDiff
+	contractPairs := utils.Flatten(diff.DeployedContracts, diff.ReplacedClasses)
```

Comment:
I don't think this is correct. `DeployedContracts` and `ReplacedClasses` state may or may not be modified. Based on the current implementation you may miss some contracts whose state was changed.

What should be done is we need to range through the `StateDiff.StorageDiffs` and `StateDiff.Nonce` and determine all the contract address whose state was modified. It is possible that a contract storage was modified but not its nonce and vice versa.

I will push a fix.

Reply:
Done